indexd/index/drivers/single_table_alchemy.py

```diff
@@ -0,0 +1,1522 @@
+            )
+
+        versioned = (
+            versioned.lower() in ["true", "t", "yes", "y"] if versioned else None
+        )
+        with self.session as session:
+            query = session.query(Record.guid, Record.urls, Record.rev)
+
+            query = query.filter(
+                func.jsonb_path_exists(
+                    Record.url_metadata, f'$.* ? (@.{key} == "{value}")'
+                )
+            )
+
+            # add version filter if versioned is not None
+            if versioned is True:  # retrieve only those with a version number
+                query = query.filter(Record.version.isnot(None))
+            elif versioned is False:  # retrieve only those without a version number
+                query = query.filter(~Record.version.isnot(None))
+
+            if url:
+                query = query.filter(
+                    func.array_to_string(Record.urls, ",").contains(url)
+                )
+            # [('did', 'url', 'rev')]
+            record_list = (
+                query.order_by(Record.guid.asc()).offset(offset).limit(limit).all()
+            )
+        return self._format_response(fields, record_list)
+
+    @staticmethod
+    def _format_response(requested_fields, record_list):
+        """loops through the query result and removes undesired columns and converts result of urls string_agg to list
+        Args:
+            requested_fields (str): comma separated list of fields to return, if not specified return all fields
+            record_list (list(tuple]): must be of the form [(did, urls, rev)], rev is not required for urls query
+        Returns:
+            list[dict]: list of response dicts
+        """
+        result = []
+        provided_fields_dict = {k: 1 for k in requested_fields.split(",")}
+        for record in record_list:
+            resp_dict = {}
+            if provided_fields_dict.get("did"):
+                resp_dict["did"] = record[0]
+            if provided_fields_dict.get("urls"):
+                resp_dict["urls"] = record[1] if record[1] else []
+
+            # check if record is returned in tuple
+            if provided_fields_dict.get("rev") and len(record) == 3:
+                resp_dict["rev"] = record[2]
+            result.append(resp_dict)
+        return result
+
+
+def check_url_metadata(url_metadata, record):
+    """
+    create url metadata record in database
+    """
+    urls = {u for u in record.urls}
+    for url, metadata in url_metadata.items():
+        if url not in urls:
+            raise UserError("url {} in url_metadata does not exist".format(url))
+    return url_metadata
+
+
+def get_record_if_exists(did, session):
+    """
+    Searches for a record with this did and returns it.
+    If no record found, returns None.
+    """
+    return session.query(Record).filter(Record.guid == did).first()
+
+
+SCHEMA_MIGRATION_FUNCTIONS = []
+CURRENT_SCHEMA_VERSION = len(SCHEMA_MIGRATION_FUNCTIONS)
```

Comment:
does something break if you remove these?

Reply:
removed